src/commands/containers/browseContainer.ts

```diff
@@ -0,0 +1,114 @@
+import { captureCancelStep } from '../../utils/captureCancelStep';
+
+type BrowseTelemetryProperties = TelemetryProperties & { possiblePorts?: number[], selectedPort?: number };
+
+type ConfigureBrowseCancelStep = 'node' | 'port';
+
+async function captureBrowseCancelStep<T>(cancelStep: ConfigureBrowseCancelStep, properties: BrowseTelemetryProperties, prompt: () => Promise<T>): Promise<T> {
+    return await captureCancelStep(cancelStep, properties, prompt)
+}
+
+// NOTE: These ports are ordered in order of preference.
+const commonWebPorts = [
+    443,    // SSL
+    80,     // HTTP
+    3000,   // (Node.js) Express.js
+    3001,   // (Node.js) Sails.js
+    5001,   // (.NET Core) ASP.NET SSL
+    5000,   // (.NET Core) ASP.NET HTTP
+    8080,   // (Node.js)
+    8081    // (Node.js)
+];
+
+const commonSslPorts = [
+    443,    // SSL
+    5001    // (.NET Core) ASP.NET SSL
+];
+
+function parsePortAndProtocol(portAndProtocol: string): { port: number, protocol: string | undefined } | undefined {
+    const splitPortAndProtocol = portAndProtocol.split('/');
+
+    if (splitPortAndProtocol.length > 0) {
+        const port = Number.parseInt(splitPortAndProtocol[0], 10);
+
+        if (!Number.isNaN(port)) {
+            return {
+                port,
+                protocol: splitPortAndProtocol.length > 1 ? splitPortAndProtocol[1].toLowerCase() : undefined
+            };
+        }
+    }
+
+    return undefined;
+}
+
+export async function browseContainer(context: IActionContext, node?: ContainerTreeItem): Promise<void> {
+    const telemetryProperties = <BrowseTelemetryProperties>context.telemetry.properties;
+
+    if (!node) {
+        node = await captureBrowseCancelStep('node', telemetryProperties, () => ext.containersTree.showTreeItemPicker<ContainerTreeItem>(ContainerTreeItem.allContextRegExp, context));
+    }
+
+    const inspectInfo = await node.getContainer().inspect();
+
+    const ports = inspectInfo.NetworkSettings && inspectInfo.NetworkSettings.Ports || {};
+    const possiblePorts =
+        Object.keys(ports)
+              .map(portAndProtocol => ({ mappings: ports[portAndProtocol], containerPort: parsePortAndProtocol(portAndProtocol) }))
+              // Exclude port designations we cannot recognized...
+              .filter(port => port.containerPort !== undefined)
+              // Exclude ports that are non-TCP-based (which should be the default)...
+              .filter(port => port.containerPort.protocol === undefined || port.containerPort.protocol === 'tcp')
+              // Exclude ports not mapped to the host...
+              .filter(port => port.mappings && port.mappings.length > 0);
+
+    telemetryProperties.possiblePorts = possiblePorts.map(port => port.containerPort.port);
+
+    if (possiblePorts.length === 0) {
+        // tslint:disable-next-line: no-floating-promises
+        ext.ui.showWarningMessage('No valid ports were mapped from the container to the host.');
```

Comment:
Do we want to leave a cancel step in telemetry here, and/or throw a `UserCancelledError`?

Reply:
I tried several variants and this seemed to work best (or was the least worst).  `UserCancelledError` doesn't seem technically correct, as the user didn't make a choice--the container just didn't have any ports to browse to.  Throwing any other kind of error results in the error dialog with the "report issue" button, which we clearly don't want.  You might argue that the command result should be `Failed` in this case, but I'm not sure it matters from the telemetry analysis point of view.  We can pick out this scenario by looking for a completed command, where the possible ports is the empty set and no port was selected.

Reply:
For the somewhat similar case where the Azure registry already exists (the one Paul hit), I had it throw a `UserCancelledError` as a way of getting out of the execution; so it wouldn't be unprecedented.